python_modules/dagster/dagster/_utils/caching_instance_queryer.py

```diff
@@ -383,3 +396,19 @@ def get_current_minutes_late_for_key(
             used_data_times=used_data_times,
             available_data_times={key: evaluation_time for key in used_data_times},
         )
+
+
+def _extract_tags(record: Optional[EventLogRecord]) -> Mapping[str, str]:
+    if record is None:
+        return frozendict()
+    dagster_event = record.event_log_entry.dagster_event
+    if dagster_event is None:
+        raise DagsterInvariantViolationError(
+            "Attempted to extract tags from record with no dagster event."
+        )
+    if dagster_event.event_type_value != DagsterEventType.ASSET_MATERIALIZATION:
+        raise DagsterInvariantViolationError(
+            "Attempted to extract tags from record with dagster event of type "
+            f"{dagster_event.event_type_value}."
+        )
+    return frozendict(dagster_event.step_materialization_data.materialization.tags or {})
```

Comment:
Why use `frozendict`? Can we rely on the type system to enforce rather than doing a copy?

Reply:
this is because every argument passed into a cached_method needs to be hashable

Reply:
Ahh, got it. Fine in that case.